File: Python3/프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    # 방향 벡터 (상, 우, 하, 좌)
    dx = [-1, 0, 1, 0]
    dy = [0, 1, 0, -1]

    result = 0
    height = len(maps)  # 맵의 높이
    width = len(maps[0])  # 맵의 너비

    # 방문 여부를 저장할 2차원 리스트
    visited = [[0] * width for _ in range(height)]
    # 거리 정보를 저장할 2차원 리스트
    distance = [[0] * width for _ in range(height)]
    # BFS에 사용할 큐
    queue = deque()

    # 시작점 초기화
    visited[0][0] = 1
    distance[0][0] = 1
    queue.append((0, 0))

    while queue:
        now = queue.popleft()  # 큐의 앞에서 현재 노드를 꺼냄

        for i in range(4):  # 4개의 방향을 탐색
            x = now[0] + dx[i]
            y = now[1] + dy[i]

            # 맵의 범위를 벗어나지 않는지 확인
            if 0 <= x <= height - 1 and 0 <= y <= width - 1:
                # 방문하지 않았고 길이 있는 경우
                if visited[x][y] == 0 and maps[x][y] == 1:
                    visited[x][y] = 1  # 방문 표시
                    distance[x][y] = distance[now[0]][now[1]] + 1  # 거리 갱신
                    queue.append((x, y))  # 다음 좌표 큐에 추가
                    
    result = distance[height - 1][width - 1]
    
    # 도달할 수 없을 때
    if result == 0:
        result = -1

    return result  

File: C++/프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include <vector>
#include <queue>

using namespace std;

// 상, 하, 좌, 우로 움직이기 위한 방향 벡터
int dx[] = {-1, 1, 0, 0};
int dy[] = {0, 0, -1, 1};

int solution(vector<vector<int>> maps) {
    int n = maps.size();       // 맵의 행 길이
    int m = maps[0].size();    // 맵의 열 길이
    
    // 방문 여부를 저장할 벡터
    vector<vector<bool>> visited(n, vector<bool>(m, false));
    
    // BFS를 위한 큐 (x, y, 현재까지의 거리)
    queue<pair<int, int>> q;
    queue<int> distances;
    
    // 시작점(0, 0)을 큐에 넣고 방문 처리
    q.push({0, 0});
    distances.push(1);  // 시작점도 거리에 포함되므로 1부터 시작
    visited[0][0] = true;
    
    // BFS 탐색
    while (!q.empty()) {
        int x = q.front().first;
        int y = q.front().second;
        int dist = distances.front();
        q.pop();
        distances.pop();
        
        // 목적지에 도착한 경우, 현재까지의 거리를 반환
        if (x == n - 1 && y == m - 1) {
            return dist;
        }
        
        // 상, 하, 좌, 우로 이동
        for (int i = 0; i < 4; i++) {
            int nx = x + dx[i];
            int ny = y + dy[i];
            
            // 맵의 경계를 벗어나지 않고, 방문하지 않았으며, 갈 수 있는 길(1)인 경우
            if (nx >= 0 && nx < n && ny >= 0 && ny < m && !visited[nx][ny] && maps[nx][ny] == 1) {
                q.push({nx, ny});
                distances.push(dist + 1);  // 거리를 1 증가
                visited[nx][ny] = true;    // 방문 처리
            }
        }
    }
    
    // 목적지에 도착하지 못한 경우 -1 반환
    return -1;
}
